internal/job/job.go

```diff
@@ -0,0 +1,93 @@
+package job
+
+import (
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/job/store"
+	"github.com/hashicorp/boundary/internal/oplog"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	// DefaultJobTableName is the default name of the database table that is used to persist a Job
+	DefaultJobTableName = "job"
+)
+
+// Job represents work that should be run at a predetermined time and needs to be synchronized
+// across servers to ensure that one and only one instance of a job is running at any given time.
+type Job struct {
+	*store.Job
+	tableName string `gorm:"-"`
+}
+
+// NewJob creates a new in memory Job.
+//
+// • Name is the human-friendly name of the job.
+//
+// • Code is not user facing and should be used to distinguish unique jobs of
+// the same type that can run in parallel.
+//
+// • Description is the human-friendly description of the job.
+//
+// WithNextScheduledRun() is the only valid option.  If this option is not
+// provided the NextScheduledRun of the job will default to zero time, and be available
+// to run immediately.
+func NewJob(name, code, description string, opt ...Option) (*Job, error) {
+	const op = "job.NewJob"
+	if name == "" {
+		return nil, errors.New(errors.InvalidParameter, op, "missing name")
+	}
+	if code == "" {
+		return nil, errors.New(errors.InvalidParameter, op, "missing code")
+	}
+	if description == "" {
+		return nil, errors.New(errors.InvalidParameter, op, "missing description")
+	}
+
+	opts := getOpts(opt...)
+	job := &Job{
+		Job: &store.Job{
+			Name:             name,
+			Description:      description,
+			Code:             code,
+			NextScheduledRun: opts.withNextScheduledRun,
```

Comment:
Yeah I saw that in your latest PR.  I was originally going down this route, but then changed it to having each job determine the time to run the next job based on a time input which would be the database current_time. `NextRun(currentTime time.Time) time.Time`.  This was done to allow for the next run to be calculated based on cron expression or similar rather than just every 15 minutes.

However as I am coding I am not loving this as much and thinking of just limiting jobs to returning a durtaion for now, lets chat Monday